### PR TITLE
Add "TIES w\ DARE", ModelStock and Geometric Median (Original!)

### PR DESCRIPTION
**Pull request by 6DammK9**
Self explained. I decided to put the condition as `vote_sgn > 0.0` to bypass type conversion. ~~Meanwhile it looks like a ReLU.~~
May use it in next version of AstolfoMix for comparasion.

### CODE REVIEW

**Comment by 6DammK9** [comment]
Found that the branch will be sync in auto. TIES w/ DARE has been added. Hyperparater is under searching. Hope that it will work.

**Comment by ljleb** [comment]
is this ready to be reviewed/merged?

**Comment by 6DammK9** [comment]
> is this ready to be reviewed/merged?

Sure. The algorithm implementation is done for me.

File: sd_mecha/merge_methods.py [hunk 1]
```diff
@@ -608,11 +680,95 @@ def overlapping_sets_pmf(n, p, overlap, overlap_emphasis):
...
+
+        objective_value = geometric_median_objective(median, models, weights)
+        if abs(prev_obj_value - objective_value) <= ftol * objective_value:
+            break
+        
+    return weighted_average(models, new_weights)
+
+def weighted_average(points, weights):
+    # weighted_average_component is not even required.
+    return torch.sum(torch.stack([p * weights[i] for i, p in enumerate(points)]), dim=0) / weights.sum()
```
Comment (ljleb) [inline sd_mecha/merge_methods.py:768]: Is it accurate if I say that this could take a lot of memory given it requires the same key of all models to be loaded at the same time?

If that is the case, then is it possible to implement the geometric median in an iterated manner? If not, then I guess there is no way around this.

File: sd_mecha/merge_methods.py [hunk 2]
```diff
@@ -133,15 +141,26 @@ def add_cosine_generic(a: Tensor, b: Tensor, alpha: float, similarity: Tensor) -
...
 def ties_sum(  # aka add_difference_ties
     *models: Tensor | LiftFlag[MergeSpace.DELTA],
     k: Hyper = 0.2,
+    vote_sgn: Hyper = 0.0,
+    apply_stock: Hyper = 0.0,
+    cos_eps: Hyper = 1e-6,
+    apply_median: Hyper = 0.0,
+    eps: Hyper = 1e-6,    
+    maxiter: Hyper = 100, 
+    ftol: Hyper =1e-20,
```
Comment (ljleb) [inline sd_mecha/merge_methods.py:163]: One thing to consider here is recipes that are already serialized and existing user code. If there are any comfyui workflows that use `ties_sum` from comfy-mecha for example, then the existing nodes will not see these new inputs added to the node. This is not too much a problem as long as the default values would preserve the original behavior without the parameters. Can you confirm this is the case?

As a side note, I am generally in favor of writing new merge methods instead of adding parameters to existing ones. Do you think it is possible to put the new code in a separate fundamental merge method? I wonder how practical that would be. All of the new code, except for the `vote_sgn` parameter, seems to be located at the end of the method, so maybe it is? Let me know if you think it is possible.

**Review by ljleb — COMMENTED** [review]
Good work overall!

If we could remove or maybe reformat some of the comments to be less temporal, then when we are done with the comments above I have left, I think I would consider this good enough for merging.

File: sd_mecha/merge_methods.py (hunk 1 above)
Comment (6DammK9) [reply sd_mecha/merge_methods.py:768]: - For the [pythonic list comprehension](https://www.w3schools.com/python/python_lists_comprehension.asp), yes. Nice parallelization with $O(N)$ of memory usage. 1.216TB for 116 SDXL under fp64. Notice that 116 SDXL in fp16 is already 750GB, which is 3TB for fp64, therefore I think 40% of overall models is already memory efficient. The $O(N)$ part will be explained seperately.
- From [wiki](https://en.wikipedia.org/wiki/Geometric_median),  Weiszfeld's algorithm is already an iterlative algorithm:
> One common approach of this type, called Weiszfeld's algorithm after the work of [Endre Weiszfeld](https://en.wikipedia.org/wiki/Endre_Weiszfeld),[[16]](https://en.wikipedia.org/wiki/Geometric_median#cite_note-16) is a form of [iteratively re-weighted least squares](https://en.wikipedia.org/wiki/Iteratively_re-weighted_least_squares).
...
This method converges for almost all initial positions
- Since geometric median is NOT statistics median, it cannot be an $O(1)$ algorithm like the [n_average](https://github.com/ljleb/sd-mecha/blob/main/examples/n_average.py). A complete set of "weighted average" are required to calculate the L2 distance for all the existing models, and I believe that it is already done inplace. Strict `list.append()` approach may reduce memory footprint to $O(1)$, but it will take at least 40x time to iterlate, which will surely [exceed my implemented test case](https://github.com/6DammK9/sd-mecha/blob/main/test/unit_test_geom_median.py#L78), and my "1.216TB" merge already used 21 hours, 40x of them will spend more than 1 month to do so. Meanwhile merging a few models with this algorithm will make this algorithm sounds ineffective due to [lack of confidence](https://www.calculator.net/sample-size-calculator.html).
- For 3 models under fp16, it will probably take 30 minutes and 30GB of RAM, which is not easy either. Moving it to `cuda` doesn't help either, because it already converges and early-stopped with around 6-10 iterlations. You can try with the [test case](https://github.com/6DammK9/sd-mecha/blob/main/test/unit_test_geom_median.py#L78),  
- BTW TGMD merge (modified TIES / DARE + GeometricMedian) will not slower, and the figures above are already for TGMD. Plain GM will be 1x in time with 0.75x of memory requiement.
>  If not, then I guess there is no way around this.
- I don't know what you mean. Hope that you're not concerning about hardware requirement because most ML algorithm are $O(N^3)$ in complexity, and parameter size of SDXL model is huge (3.6b) for any algorithm, otherwise we are limited with algebra. This is still $O(N)$ like TIES / DARE / SLERP / ModelStock, not even close to rebasin, which is $O(NlogN)$ in time. 
- Also, I expect such merging is being done solely in CPU, with an average workstation, like X99 + 256GB of RAM, which should support for 100 SDXL models in fp16.

File: sd_mecha/merge_methods.py (hunk 2 above)
Comment (6DammK9) [reply sd_mecha/merge_methods.py:163]: > In favor of writing new merge methods

I have made a [research log](https://github.com/6DammK9/nai-anime-pure-negative-prompt/blob/main/ch05/README_XL.MD#findings-on-sdxl-ties) for such SDXL merge. The following suggestion (split to 3 algos with no stacking) is up to you:
- First, make a set of **original** TIES merge, then "TIES w/ DARE" as DARE, and try to figure out how to implement DARE on top of a known receipt. It may work for some (or most?) cases as [implemented in other repos](https://github.com/54rt1n/ComfyUI-DareMerge/blob/master/components/dare.py#L15), but it is not enough for me for my [Pony x SDXL merge](https://civitai.com/models/309514?modelVersionId=484834). It can be just a monolithic method with union of parameters.
- Then, obviously Geometric merge can be implemented alone, and I expect it behaves like `weighted_sum` but closer to base model.
- Finally implement this TGMD merge **exclusively**. Since I have already [overrided more then half of the original algorithm](https://github.com/6DammK9/nai-anime-pure-negative-prompt/blob/main/ch01/modelstock.md#spinoff-model-stonk-by-calculate-geometric-median-with-gradient-descent), I think it is suitable to trim all the parameters out, leaving only `p=0.1` for the dropout part. Even `maxiter` / `eps` / `ftol` are _safe to hide from user's sight._ Even `k=1.0` / `alpha=1.0` are in the same case: They _defines_ TGMD merge as an algorithm.

Since I'm the minority who [use dedicated python notebook](https://github.com/6DammK9/nai-anime-pure-negative-prompt/blob/main/ch05/astolfo_mix_e2e.ipynb) and a [ridiculous workstation build](https://github.com/6DammK9/nai-anime-pure-negative-prompt/blob/main/ch04/ice_lake_ws.md), I always able to find a way to keep my TGMD merge after breaking changes accoring to ComfyUI adaption, so feel free to make your decision. To get users well informed, we can specify that such memory intensive (O(N) instead of O(1), which is still harsh to me) algoithm will be forced to use `device='cpu'` to prevent OOM.

**Comment by 6DammK9** [comment]
> code comments or maybe reformat them 

Maybe [Linting](https://code.visualstudio.com/docs/python/linting) with [LaTeX](https://oeis.org/wiki/List_of_LaTeX_mathematical_symbols) variable names. However I still favour [list comprehension](https://www.w3schools.com/python/python_lists_comprehension.asp) because [lambda expression](https://www.geeksforgeeks.org/python-lambda-anonymous-functions-filter-map-reduce/) is highly correlated to algorithm expression itself and being pythonic, meanwhile `torch` has already optimized in C layer, even faster than `numpy` / `scipy` / `sklearn` / `networkx` etc.
If the code is short and clean enough, even dropping the comment is feasible, like the Weiszfeld's algorithm alone.

File: sd_mecha/merge_methods.py (hunk 1 above)
Comment (ljleb) [reply sd_mecha/merge_methods.py:768]: > If not, then I guess there is no way around this.
By this I mean that the current implementation might be as optimal as we can make it to be, since there is no known way to make it take less memory while preserving the speed of execution.

Comment (ljleb) [reply sd_mecha/merge_methods.py:768]: Thanks for the clarification, I see why this parallel approach was taken now.

File: sd_mecha/merge_methods.py (hunk 2 above)
Comment (ljleb) [reply sd_mecha/merge_methods.py:163]: > Since I'm the minority who [use dedicated python notebook](https://github.com/6DammK9/nai-anime-pure-negative-prompt/blob/main/ch05/astolfo_mix_e2e.ipynb) and a [ridiculous workstation build](https://github.com/6DammK9/nai-anime-pure-negative-prompt/blob/main/ch04/ice_lake_ws.md), I always able to find a way to keep my TGMD merge after breaking changes accoring to ComfyUI adaption, so feel free to make your decision.

Got it. I appreciate the cordial and constructive approach to the conversation. I'll try to break the code into smaller recipes and then we can merge.

**Comment by ljleb** [comment]
The remote branch name is `main` and I am having trouble pushing to it as a result. For this reason, I am moving the work over to #37.
